Lib/test/test_asyncio/test_taskgroups.py

```diff
@@ -997,6 +997,94 @@ class MyKeyboardInterrupt(KeyboardInterrupt):
         self.assertIsNotNone(exc)
         self.assertListEqual(gc.get_referrers(exc), no_other_refs())
 
+    async def test_taskgroup_cancel_children(self):
```

Comment:
can you add a test that a race function works, eg there's only one winner

```python
async def test_race_one_winner():

    async def race(*fns):
        outcome = None

        async def run(fn):
            nonlocal outcome
            outcome = await fn()
            tg.stop()

        async with asyncio.TaskGroup() as tg:
            for fn in fns:
                tg.create_task(run(fn))

    event = asyncio.Event()
    record = []
    async def fn_1():
        await event.wait()
        record.append(1)
        return 1

    async def fn_2():
        await event.wait()
        record.append(2)
        return 2

    async def trigger_event():
        event.set()
        await asyncio.sleep(10)

    outcome = await race(fn_1, fn_2, trigger_event)
    self.assertEquals(outcome, 1)
    self.assertListEquals(record, [1])
```

Reply:
I'm not sure we should expect only one winner, and vaguely recall Trio guidance against such expectations.  I'm not sure such a guarantee is useful in practice, because a task wouldn't cancel a task group until its real work was completed, and there is no way to prevent multiple tasks finishing their work on the same scheduler pass (short of employing locks within the tasks).

Would you clarify your expectation?  For example, "for any tasks transitively under a TaskGroup that may call `tg.cancel()`, only one such task is able to do so".

Reply:
If they do finish on the same scheduler count only one will resume, so it can call .cancel() on sibling tasks to prevent them finishing

This behaviour is already required by staggered_race